### PR TITLE
feat(tier4_autoware_utils): add new calcSignedArclength Function

### PR DESCRIPTION
**Pull request by purewater0901**
Signed-off-by: yutaka <purewater0901@gmail.com>

## Description
Add new calc signed arc function which has distance and yaw threshold. 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by kenji-miyake — COMMENTED** [review]
I'll leave the review to @takayuki5168 -san!

File: common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/trajectory.hpp [hunk 1]
```diff
@@ -269,6 +269,29 @@ double calcSignedArcLength(
   return signed_length_on_traj - signed_length_src_offset;
 }
 
+/**
+ * @brief calcSignedArcLength from point to index with maximum distance and yaw threshold
+ */
+template <class T>
+boost::optional<double> calcSignedArcLength(
+  const T & points, const geometry_msgs::msg::Pose & src_pose, const size_t & dst_idx,
```
Comment (takayuki5168) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/trajectory.hpp:277]: reference is not required for size_t

Comment (purewater0901) [reply common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/trajectory.hpp]: done

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Pull request merged**